sql/src/main/java/io/crate/operation/collect/MapSideDataCollectOperation.java

```diff
@@ -463,44 +470,49 @@ private int numShards(CollectPhase collectNode, String localNodeId) {
         }
 
         // start the projection
         projectorChain.startProjections(jobCollectContext);
         try {
             LOGGER.trace("starting {} shardCollectors...", numShards);
-            return runCollectThreaded(collectNode, shardCollectors);
-        } catch (RejectedExecutionException e) {
+            return runCollectThreaded(collectNode, shardCollectors, jobCollectContext);
+        } catch (RejectedExecutionException | InterruptedException | ExecutionException e) {
             // on distributing collects the merge nodes need to be informed about the failure
             // so they can clean up their context
             // in order to fire the failure we need to add the operation directly as an upstream to get a handle
             downstream.registerUpstream(this).fail(e);
-            return Futures.immediateFailedFuture(e);
+            jobCollectContext.closeDueToFailure(e);
         }
-
+        return EMPTY_LIST;
     }
 
-    private ListenableFuture<List<Void>> runCollectThreaded(CollectPhase collectNode,
-                                                            final List<CrateCollector> shardCollectors) throws RejectedExecutionException {
+    private Collection<CrateCollector> runCollectThreaded(CollectPhase collectNode,
+                                                          final List<CrateCollector> shardCollectors,
+                                                          final JobCollectContext jobCollectContext)
+            throws RejectedExecutionException, ExecutionException, InterruptedException {
         if (collectNode.maxRowGranularity() == RowGranularity.SHARD) {
             // run sequential to prevent sys.shards queries from using too many threads
             // and overflowing the threadpool queues
-            return listeningExecutorService.submit(new Callable<List<Void>>() {
+            return listeningExecutorService.submit(new Callable<Collection<CrateCollector>>() {
                 @Override
-                public List<Void> call() throws Exception {
+                public Collection<CrateCollector> call() throws Exception {
                     for (CrateCollector collector : shardCollectors) {
                         collector.doCollect();
                     }
-                    return ONE_LIST;
+                    jobCollectContext.close();
+                    return shardCollectors;
                 }
-            });
+            }).get();
         } else {
-            return ThreadPools.runWithAvailableThreads(
-                    executor,
-                    poolSize,
-                    collectors2Callables(shardCollectors),
-                    new VoidFunction<List<Void>>());
+             ThreadPools.runWithAvailableThreads(
+                     executor,
+                     poolSize,
+                     collectors2Callables(shardCollectors),
+                     new VoidFunction<List<Void>>());
+            jobCollectContext.close();
```

Comment:
I don't think that it's correct to close the context here while the collectors are still running